### PR TITLE
bpo-38943: Fix IDLE autocomplete window not always appearing

### PR DESCRIPTION
**Pull request by JohnnyNajera**
https://bugs.python.org/issue38943

In some systems the function winconfig_event is called in a context in which wm_geometry has no effect.

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@JohnnyNajera

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Review by JohnnyNajera — COMMENTED** [review]
In some systems the function winconfig_event is called in a context in which wm_geometry has no effect. Calling it through widget.after(..) solves the problem on the systems I've checked (Ubuntu 19.04 to Ubuntu 19.10), without affecting its behavior on older systems (Tested on: Ubuntu 16.04, Ubuntu 18.04).

Furthermore, the escape key never make it to the function keypress_event, without specifically binding it.

File: Lib/idlelib/autocomplete_w.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 # before the default specific IDLE function
 KEYPRESS_SEQUENCES = ("<Key>", "<Key-BackSpace>", "<Key-Return>", "<Key-Tab>",
                       "<Key-Up>", "<Key-Down>", "<Key-Home>", "<Key-End>",
-                      "<Key-Prior>", "<Key-Next>")
+                      "<Key-Prior>", "<Key-Next>", "<Key-Escape>")
```
Comment (terryjreedy) [inline Lib/idlelib/autocomplete_w.py]: I agree that Escape should close the window, but this is a different and unrelated change that belongs on a separate issue/PR.  Revert.
```suggestion
                      "<Key-Prior>", "<Key-Next>")
```

File: Lib/idlelib/autocomplete_w.py [hunk 2]
```diff
@@ -256,7 +256,7 @@ def winconfig_event(self, event):
         else:
             # place acw above current line
             new_y -= acw_height
-        acw.wm_geometry("+%d+%d" % (new_x, new_y))
+        self.widget.after(1, lambda: self._update_geometry(acw, new_x, new_y))
```
Comment (terryjreedy) [inline Lib/idlelib/autocomplete_w.py]: The call could be delayed without adding two layers of wrapping.
```suggestion
        self.widget.after(1, acw.wm_geometry, "+%d+%d" % (x, y)) 
```
and delete the added _update_geometry.  But I don't think this is the right fix.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Without knowing what the problem to be fixed it, I cannot judge whether the 1 millisecond delay should really fix it.  I worry that calling geometry after is_configuring is set false could under some circumstances trigger another call.

My deeper problem with the patch is that it is yet another kludge on what I consider a flawed design.  We want the positioning function to be done once.  Yet it is bound to an event that happens multiple times.  The first Configure event is from 'wm_geometry("+10000+10000")' in show_window.  This is not needed (and would not work on a future 16K monitor) and is done for no other popup  (window.withdraw also makes a window invisible).  That event is later bound to winconfig_event and we depend on that post-event binding to trigger the function.  Instead, we should just call the positioning function when needed, as we do with calltips and all dialogs.

Within the function, 'acw.update' triggers more Configure events (4 on Windows).  So is_configuring is needed to avoid infinite recursion.  On Windows, there are other Configure events, so the function is unbound from Configure events in the first call.  Without all this nonsense, we don't seem to have no-show problems on other systems.

If we touch this stuff, we have to manually test on all three systems, so I want to refactor to make it better.  Perhaps we should make this a tooltip subclass, as with calltips.

File: Lib/idlelib/autocomplete_w.py (hunk 1 above)
Comment (JohnnyNajera) [reply Lib/idlelib/autocomplete_w.py]: OK. Should I open a new one with just this change?

File: Lib/idlelib/autocomplete_w.py (hunk 2 above)
Comment (JohnnyNajera) [reply Lib/idlelib/autocomplete_w.py]: Personally I think the lambda approach is more readable, but fair enough.

**Comment by JohnnyNajera** [comment]
Hi,
I agree the entire design could be rewritten.

Regarding the 1-milisecond - It's not about 1-milisecond, but about adding the function to the mainloop queue, so it is called in a context in which it works.
I agree this is indeed a quick solution to a an entire class that can be rewritten, but in the meanwhile I find myself manually patching this on every Ubuntu 19.10 I'm working on.
Maybe we can merge this, while I will gladly work on a redesign?

I have made the requested changes; please review again.

P.S.
Regarding the `update` function, how about calling `update_idletasks` instead?

**Comment by terryjreedy** [comment]
I appears that you have not yet signed the CLA.  Please do so.

File: Lib/idlelib/autocomplete_w.py [hunk 3]
```diff
@@ -256,7 +256,7 @@ def winconfig_event(self, event):
         else:
             # place acw above current line
             new_y -= acw_height
-        acw.wm_geometry("+%d+%d" % (new_x, new_y))
+        self.widget.after(1, acw.wm_geometry, "+%d+%d" % (x, y))
```
Comment (taleinat) [inline Lib/idlelib/autocomplete_w.py]: This should use `new_x` and `new_y`, rather than `x` and `y`.

**Comment by taleinat** [comment]
> Maybe we can merge this, while I will gladly work on a redesign?

If we can't easily find the root cause for this issue, I tend to agree.

I'll try testing on my Ubuntu 18.10 machine to see if this reproduces. (Ideally we'd find a better fix for this which doesn't require a rewrite.)

**Review by taleinat — DISMISSED** [review]

**Comment by taleinat** [comment]
>  Regarding the `update` function, how about calling `update_idletasks` instead?

Assuming you are referring to the code in the same function: It requires that the auto-complete window to be fully positioned in order to query its geometry, and calling `update()` is the only reliable way to ensure that, AFAIK.

File: Lib/idlelib/autocomplete_w.py (hunk 3 above)
Comment (JohnnyNajera) [reply Lib/idlelib/autocomplete_w.py]: Of course, thank you. This happened in the last change suggestion, I didn't notice.

**Comment by JohnnyNajera** [comment]
A different solution can be just calling `update_idletasks` right after `wm_geometry` which works just as well, which might be slightly preferred aesthetically.
BTW, any idea why my CLA sign is not approved? It's been two US work days.
I have made the requested changes; please review again

**Comment by taleinat** [comment]
> A different solution can be just calling `update_idletasks` right after `wm_geometry` which works just as well, which might be slightly preferred aesthetically.

That seems like a more appropriate solution to me. (And the difference isn't just aesthetic.)

**Comment by JohnnyNajera** [comment]
> (And the difference isn't just aesthetic.)

It is in the sense we don't quite understand both of them. But fair enough, I have committed it.

I have made the requested changes; please review again.

**Comment by JohnnyNajera** [comment]
What do you say? should we merge this? @taleinat

**Review by terryjreedy — APPROVED** [review]

**Comment by terryjreedy** [comment]
The tooltip event handler, used for calltips, uses .update_idletasks and seems to be working fine.  I check all 3 current completions both auto and manual on windows and all continue to work.  So I am comfortable with this -- and will check on Mac Mohave when the .rcs are released tomorrow or so.
